google/cloud/pubsub/internal/tracing_message_batch.cc

```diff
@@ -35,66 +35,83 @@ void TracingMessageBatch::SaveMessage(pubsub::Message m) {
-    std::lock_guard<std::mutex> lk(mu_);
+    std::lock_guard<std::mutex> lk(message_mu_);
     message_spans_.push_back(std::move(active_span));
   }
   child_->SaveMessage(std::move(m));
 }
 
-void TracingMessageBatch::AddMessageSpanMetadata() {
-  for (auto& message_span : message_spans_) {
-    message_span->AddEvent("gl-cpp.batch_flushed");
-  }
-}
+namespace {
 
-void TracingMessageBatch::Flush() {
+opentelemetry::nostd::shared_ptr<opentelemetry::trace::Span> MakeParentSpan(
+    std::vector<
+        opentelemetry::nostd::shared_ptr<opentelemetry::trace::Span>> const&
+        message_spans) {
   using opentelemetry::trace::SpanContext;
   using AttributesList =
       std::vector<std::pair<opentelemetry::nostd::string_view,
                             opentelemetry::common::AttributeValue>>;
   auto constexpr kMaxOtelLinks = 128;
   std::vector<std::pair<SpanContext, AttributesList>> links;
-  auto batch_size = message_spans_.size();
+  auto batch_size = message_spans.size();
   links.reserve(batch_size);
-
   // If the batch size is less than the max size, add the links to a single
   // span.
   if (batch_size < kMaxOtelLinks) {
     std::transform(
-        message_spans_.begin(), message_spans_.end(), std::back_inserter(links),
+        message_spans.begin(), message_spans.end(), std::back_inserter(links),
         [i = static_cast<std::int64_t>(0)](auto const& span) mutable {
           return std::make_pair(
               span->GetContext(),
               AttributesList{{"messaging.pubsub.message.link", i++}});
         });
   }
-  auto batch_sink_span_parent =
+  auto batch_sink_parent_span =
       internal::MakeSpan("BatchSink::AsyncPublish",
                          /*attributes=*/
                          {{"messaging.pubsub.num_messages_in_batch",
                            static_cast<std::int64_t>(batch_size)}},
                          /*links*/ links);
 
-  // TODO(#12528): Handle batches larger than 128.
+  auto context = batch_sink_parent_span->GetContext();
+  auto trace_id = internal::ToString(context.trace_id());
+  auto span_id = internal::ToString(context.span_id());
+  for (auto& message_span : message_spans) {
+    message_span->AddEvent("gl-cpp.batch_flushed");
+    message_span->SetAttribute("pubsub.batch_sink.trace_id", trace_id);
+    message_span->SetAttribute("pubsub.batch_sink.span_id", span_id);
+  }
 
-  // This must be called before we clear the message spans.
-  AddMessageSpanMetadata();
-  // Clear message spans.
-  message_spans_.clear();
+  return batch_sink_parent_span;
+}
 
-  batch_sink_spans_.push_back(batch_sink_span_parent);
+}  // namespace
 
-  // Set the batch sink parent span.
+void TracingMessageBatch::Flush() {
+  decltype(message_spans_) message_spans;
+  {
+    std::lock_guard<std::mutex> lk(message_mu_);
+    message_spans.swap(message_spans_);
+  }
+
+  // TODO(#12528): Handle batches larger than 128.
+  auto batch_sink_parent_span = MakeParentSpan(std::move(message_spans));
```

Comment:
nit: `std::move()` into a function that receives by `const&` ?  Consider changing the function to receive by value, since you really don't need the `message_spans` parameter after calling the function.

Reply:
Done

Reply:
> you really don't need the `message_spans` parameter after calling the function.

I think the by-value issue is best thought about as whether `MakeParentSpan()` always copies the vector (or its contained references), and that does not seem to be the case here.  So, `message_spans` would be better passed as a `const&`, no matter what the caller decides to do with it next.